### ISSUE TITLE
Metadata: Add settings for interpreting the date of old files without Exif data

### ISSUE THREAD

**Issue by deric** [photoprism/photoprism#1691]
I'm having issues with old photos from 2005 being categorized into 2020 just because of incorrect `File Modification Date/Time`.

I guess there might not be a universal solution to suite everyone. If I understand this correctly photoprism currently uses single timestamp to represent a time when was [the photo taken](https://github.com/photoprism/photoprism/blob/bc86997ba3923498ba13a9e176dfa55ab26ed3f3/internal/meta/exif.go#L241):

```go
	if value, ok := tags["DateTimeOriginal"]; ok && ValidDateTime(value) {
		takenAt = value
	} else if value, ok := tags["DateTimeDigitized"]; ok && ValidDateTime(value) {
		takenAt = value
	} else if value, ok := tags["CreateDate"]; ok && ValidDateTime(value) {
		takenAt = value
	} else if value, ok := tags["DateTime"]; ok && ValidDateTime(value) {
		takenAt = value
	}
```
For some photo collections it might be useful to work either with EXIF's `DateTimeDigitized` or `CreateDate` etc. More robust solution could work with "primary date time attribute" and "fallback date time attributes".

**Comment by lastzero** [comment]
Do you have Exiftool disabled? Otherwise, this is what already happens (of course we can also add the fallback to our regular Exif parser even if that doesn't make a difference with Exiftool enabled):

https://github.com/photoprism/photoprism/blob/develop/internal/meta/data.go#L16

**Comment by deric** [comment]
No, Exiftool is enabled. Obviously the metadata for theses photos are messed up, but photoprisim uses EXIF's `ModifyDate` or (`FileModifyDate`) in this case whilst `DateTimeOriginal` or `CreateDate` would be a better choice.

```
$ exiftool -time:all -G1 -a -s HPIM5976.JPG
[System]        FileModifyDate                  : 2020:06:10 18:56:51+02:00
[System]        FileAccessDate                  : 2021:10:29 16:05:16+02:00
[System]        FileInodeChangeDate             : 2021:08:26 19:12:08+02:00
[IFD0]          ModifyDate                      : 2020:06:10 18:56:51
[ExifIFD]       DateTimeOriginal                : 2005:09:16 09:59:31
[ExifIFD]       CreateDate                      : 2005:09:16 09:59:31
[XMP-xmp]       ModifyDate                      : 2020:06:10 18:56:51+02:00
```

**Comment by lastzero** [comment]
No it uses the modify time directly from the filesystem (which Exiftool also shows) and marks the actual year, month, and day as unknown if it is within the last 24 hours because this is unreliable (if it's an older date, it may approximately be correct so it's not set to unknown, but we could add yet another config option for this):

https://github.com/photoprism/photoprism/blob/de63227767c0c56885b55d2337f8ffcbcceae979/internal/entity/photo.go#L798 

Note that PhotoPrism will even try to get the date from file and directory names.

If you are aware of any other source or Exif field we could use, please let us know.

**Comment by lastzero** [comment]
As a developer, you can set `photo_day`, `photo_month`, and `photo_year` to `-1` for all rows that have `taken_src = ''`. This will show their date as unknown in the UI.

**Comment by deric** [comment]
I don't really care about filesystem meta data (modify or access time), I'd like to have photos sorted by the datetime when was the photo actually taken. Photos were copied from some old hdd, so the modified date could be different. Directory can be renamed, but I'd still like to see the date when was the photo taken.

**Comment by lastzero** [comment]
We also want you to see the date when was the photo taken. Is this known in your case? Is there a specific field we ignore right now and that we should add support for? You mentioned metadata is incomplete or doesn't exist...

**Comment by deric** [comment]
> We also want you to see the date when was the photo taken. Is this known in your case? 

Yes. From EXIF's `DateTimeOriginal` or `CreateDate` as shown above. But Photoprism should already be using them, right? 

> You mentioned metadata is incomplete or doesn't exist...

It's 15 years old photo. Metadata standards might have changed during that time :) But I still don't understand why Photoprism uses wrong date in this case when EXIF data is available.

Is it possible to debug this somehow? Increase log verbosity?

**Comment by lastzero** [comment]
> > We also want you to see the date when was the photo taken. Is this known in your case?
> 
> Yes. From EXIF's `DateTimeOriginal` or `CreateDate` as shown above. But Photoprism should already be using them, right?

Right. It should search for a date in the following fields:

- DateTimeOriginal
- CreationDate
- CreateDate
- MediaCreateDate
- ContentCreateDate
- DateTimeDigitized
- DateTime

> > You mentioned metadata is incomplete or doesn't exist...
> 
> It's 15 years old photo. Metadata standards might have changed during that time :) 

No, the Exif date field didn't really change. In the worst case it is missing or contains invalid characters. Did you check your event logs for Exif related warnings and errors?

https://demo.photoprism.org/library/errors

You can also enable debug mode, maybe it helps.

**Comment by deric** [comment]
```
MariaDB [photoprism]> select id, taken_at,taken_src from photos where photo_name='HPIM5976';
+-------+---------------------+-----------+
| id    | taken_at            | taken_src |
+-------+---------------------+-----------+
| 12112 | 2020-06-10 18:56:51 | meta      |
+-------+---------------------+-----------+
```

`taken_src` == `meta` means it's fetched from EXIF? or could be still from filesystem?

**Comment by lastzero** [comment]
Should be from Exif then. It checks the fields in the order I sent you. Might be needed to change that order if the actual date comes later and the wrong date comes first. Can you check that for us?

**Comment by maxammann** [comment]
I'm not sure whether this is related, but I think the photoprism is not that robust. I store the DateTimeOriginal in JPG files. Exiftool parses it as:

```console
exiftool "darktable_exported/raw0051.jpg"  "-DateTimeOriginal" -j
[{
  "SourceFile": "darktable_exported/raw0051.jpg",
  "DateTimeOriginal": "2019:08:01 00:00"
}]
```

The issue is that the seconds are missing. A valid DateTimeOriginal according to photoprism would be: `2019:08:01 00:00:00`.
Is this expected?

If this is not related then I would just open a new issue :)

**Comment by lastzero** [comment]
Unexpected values are ignored intentionally. Thanks to this, we now know about a problem that would otherwise have gone unnoticed.

Technically, according to the Exif specs, you would use spaces in places where the number isn't known. So August 1993 would be `1993:08: : : `. A timestamp like `1993:08:00 00:00` isn't valid.

Also I've expected that Exiftool normalizes incomplete or incorrectly formatted timestamps.

If you submit a feature request for incomplete timestamp support, please provide affected image files for testing 👍

**Comment by graciousgrey** [comment]
Might be the same as: https://github.com/photoprism/photoprism/issues/625
